test_util/provide_aws.py

```diff
@@ -0,0 +1,215 @@
+#!/usr/bin/env python3
+import logging
+import os
+from abc import ABCMeta, abstractmethod
+
+import boto3
+import botocore
+import retrying
+
+
+LOGGING_FORMAT = '[%(asctime)s|%(name)s|%(levelname)s]: %(message)s'
+logging.basicConfig(format=LOGGING_FORMAT, level=logging.DEBUG)
+logger = logging.getLogger(__name__)
+
+AWS_VERIFY = True
+
+
+class AwsProvision(metaclass=ABCMeta):
+    """Class for utilities shared between VPC and CF provisioning
+    """
+
+    @abstractmethod
+    def create(self):
+        pass
+
+    @abstractmethod
+    def delete(self):
+        pass
+
+    def __init__(self, region):
+        self.stack = None
+        self.stack_name = None
+        self.region = region
+        self.session = boto3.session.Session(
+            aws_access_key_id=os.environ['AWS_ACCESS_KEY_ID'],
+            aws_secret_access_key=os.environ['AWS_SECRET_ACCESS_KEY'])
+        self.ec2_client = self.session.client(
+            service_name='ec2',
+            region_name=self.region,
+            verify=AWS_VERIFY)
+
+    def set_expiration(self, expiration_string):
+        ''' For use with Mesosphere/DCOS CI
+        Instances not tagged and owned after after instantiation will be deleted
+        within 30 minutes
+        '''
+        reservation = self.ec2_client.describe_instances(Filters=[{
+                'Name': 'tag-value', 'Values': [self.stack_name]}])['Reservations']
+        for inst_i in reservation[0]['Instances']:
+            inst = self.session.resource('ec2', region_name=self.region).Instance(inst_i['InstanceId'])
+            inst.create_tags(
+                    DryRun=False,
+                    Tags=[
+                        {
+                            'Key': 'expiration',
+                            'Value': expiration_string
+                        },
+                        {
+                            'Key': 'owner',
+                            'Value': '#dcos-ci'
+                        }])
+
+    def create_key_pair(self, key_name):
+        """
+        Args:
+            key_name (str): key name to be registered in AWS
+        Returns:
+            string of contents of key
+        """
+        key = self.ec2_client.create_key_pair(KeyName=key_name)
+        return key['KeyMaterial']
+
+    def delete_key_pair(self, key_name):
+        self.ec2_client.delete_key_pair(KeyName=key_name)
+
+    @retrying.retry(stop_max_delay=1200*1000,
+                    wait_fixed=3000,
+                    retry_on_exception=lambda ex: False,
+                    retry_on_result=lambda res: res is False)
+    def wait_for_cluster_creation(self):
+        logger.info('Waiting for cluster to be created...')
+        self.stack.reload()
+        current_status = self.stack.stack_status
+        if current_status == 'CREATE_IN_PROGRESS':
+            return False
+        if current_status != 'CREATE_COMPLETE':
```

Comment:
Should detect CREATE_FAILED here (so we can exit early when we know it's dead dead), as well as hopefully do status logging or something similar to help with debugging state over time.